eliot/_action.py

```diff
@@ -808,3 +808,48 @@ def startTask(logger=None, action_type=u"", _serializers=None, **fields):
                     _serializers)
     action._start(fields)
     return action
+
+
+class TooManyCalls(Exception):
+    """
+    The callable was called more than once.
+
+    This typically indicates a coding bug: the result of
+    C{preserve_context} should only be called once, and
+    C{preserve_context} should therefore be called each time you want to
+    pass pass the callable to a thread.
+    """
+
+
+def preserve_context(f):
+    """
+    Package up the given function with the current Eliot context, and then
+    restore context and call given function when the resulting callable is
+    run. This allows continuing the action context within a different thread.
+
+    The result should only be used once.
+
+    @param f: A callable.
+
+    @return: One-time use callable that calls given function in context of
+        a child of current Eliot action.
```

Comment:
It seems like the one-time use functionality is orthogonal to the context preservation functionality.  Maybe it's convenient to bundle them together, though?  Or maybe it will be inconvenient when someone has a legitimate use-case for repeated invocation (okay, not that inconvenient, just requires calling `preserve_context` in a slightly different place).

The implementation could probably be split in two, though.  Not a big deal since this is all fairly simple stuff.  I think it would be slightly better to have the two pieces of functionality implemented as two distinct units, though.

:question: